flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java

```diff
@@ -228,7 +234,14 @@ public void startNewWorker(ResourceProfile resourceProfile) {
 
 	@Override
 	public void stopWorker(ResourceID resourceID) {
-		// TODO: Implement to stop the worker
+		ContainerId containerId = containerIdMap.remove(resourceID.getResourceIdString());
+		if (containerId != null) {
+			log.info("Stopping container {}.", containerId.toString());
+			resourceManagerClient.releaseAssignedContainer(containerId);
```

Comment:
I think in the Yarn case, the `resourceID` is generated from the stringified version of the `containerId`. Thus, you should be able to retrieve it from the provided `resourceID`.

Reply:
The ContainerId for YARN can be only generated using ContainerId.newInstance, which requires the application attempt id and the container Id. The resourceID only contains the container Id information, so it's not sufficient to rebuild the YARN ContainerId object that the releaseAssignedContainer method takes.

Reply:
Then we should change the `Worker` type to also include the Mesos `ContainerId` and pass it to the `stopWorker` method.